pkg/healthchecker/healthchecker.go

```diff
@@ -18,61 +18,152 @@ package healthchecker
-	return s.waitUntilAllReplicasAreRunning(ctx, l, logSuffix)
-}
+func (s *HealthCheckerImpl) Probe(ctx context.Context, l logr.Logger, logSuffix string, nodeID int) error {
+	l.V(int(zapcore.DebugLevel)).Info("Health check probe", "label", logSuffix, "nodeID", nodeID)
+	stsname := s.cluster.StatefulSetName()
+	stsnamespace := s.cluster.Namespace()
+	podname := fmt.Sprintf("%s-%v", stsname, nodeID)
+
+	//we sleep 1 minunte to ensure the pre-stop hook finished
+	time.Sleep(1 * time.Minute)
+	//SKIP KILL because we just restarted????
+	// err := sendKillSignal(ctx, l, logSuffix, podname, nodeID)
+	// if err != nil {
+	// 	return err
+	// }
 
-func (s *HealthCheckerImpl) checkAllReplicasAreRunning(ctx context.Context, l logr.Logger, logSuffix string) error {
-	l.V(int(zapcore.DebugLevel)).Info("checkAllReplicasAreRunning", "label", logSuffix)
-	underReplicasRangeSum, err := getUnderReplicatedRanges(ctx, s.Db, logSuffix)
+	//isready check for all pods from sts
+	err := kube.WaitUntilAllStsPodsAreReady(ctx, s.clientset, l, stsname, stsnamespace, 30*time.Minute, 10*time.Minute)
 	if err != nil {
 		return err
 	}
-	if underReplicasRangeSum != 0 {
-		return errors.New("under replica is not zero")
+
+	sts, err := s.clientset.AppsV1().StatefulSets(stsnamespace).Get(ctx, stsname, metav1.GetOptions{})
+	if err != nil {
+		return kube.HandleStsError(err, l, stsname, stsnamespace)
+	}
+	//TODO: add goroutine for each partition
+	for partition := *sts.Spec.Replicas - 1; partition >= 0; partition-- {
+		//we check the metric for all the pods
+		err = s.waitUntilUnderReplicatedMetricIsZero(ctx, l, logSuffix, podname, partition)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//sendKillSignal it is not necessary in our scenario... we just restarted the node and run drain as pre-stop hook
+func sendKillSignal(ctx context.Context, l logr.Logger, logSuffix, podname string, nodeID int) error {
+	l.V(int(zapcore.DebugLevel)).Info("sendKillSignal", "label", logSuffix, "pod", podname, "nodeID", nodeID)
+	//assuming that crdb process always start with id 1
+	process, err := os.FindProcess(1)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get process cockroach")
+	}
+	//TODO:Check if process.Kill or process.Signal usage
+	err = process.Signal(syscall.Signal(syscall.SIGKILL))
+	if err != nil {
+		return errors.Wrapf(err, "failed to get kill process cockroach")
 	}
 	return nil
 }
-func (s *HealthCheckerImpl) waitUntilAllReplicasAreRunning(ctx context.Context, l logr.Logger, logSuffix string) error {
+
+func (s *HealthCheckerImpl) waitUntilUnderReplicatedMetricIsZero(ctx context.Context, l logr.Logger, logSuffix, podname string, partition int32) error {
 	f := func() error {
-		return s.checkAllReplicasAreRunning(ctx, l, logSuffix)
+		return s.checkUnderReplicatedMetric(ctx, l, logSuffix, podname, partition)
 	}
 	b := backoff.NewExponentialBackOff()
-	b.MaxElapsedTime = 30 * time.Minute
+	b.MaxElapsedTime = 3 * time.Minute
```

Comment:
For this we just want to do a poll every 10 seconds, maybe with a bit of jitter but should be fine as we hit each node independently. If we don't get 0 for all nodes within 3 minutes then we should stop.

Reply:
this works as expected